bridge/server/user/oauth.ts

```diff
@@ -1,225 +1,86 @@
-  router.get('/login', async (req: Request, res: Response) => {
-    let authResponse;
-    try {
-      authResponse = await axios({
-        method: 'get',
-        url: authorizationEndpoint,
-      });
-    } catch (error) {
-      const err = error as AxiosError;
-      console.log(`Error while handling the login request. Cause : ${err.message}`);
-      return res.render('error', {
-        title: 'Internal error',
-        message: 'Error while handling the login request.',
-        location: getRootLocation(),
-      });
-    }
-
-    if (!authResponse.data.hasOwnProperty(AUTH_URL)) {
-      return res.render('error', {
-        title: 'Invalid state',
-        message: 'Failure to obtain login details.',
-        location: getRootLocation(),
-      });
-    }
-
-    res.redirect(authResponse.data[AUTH_URL]);
-    return res;
+  return new ssoIssuer.Client({
+    client_id: clientId,
+    ...(clientSecret && { client_secret: clientSecret }),
+    redirect_uris: [redirectUri],
+    response_types: ['code'],
+    token_endpoint_auth_method: clientSecret ? 'client_secret_basic' : 'none',
+    id_token_signed_response_alg: process.env.OAUTH_ID_TOKEN_ALG || 'RS256',
   });
+}
 
-  /**
-   * Router level middleware for redirect handling
-   */
-  router.get('/oauth/redirect', async (req: Request, res: Response) => {
-    const authCode = req.query.code;
-    const state = req.query.state;
-
-    if (authCode === undefined || state === undefined) {
-      return res.redirect(getRootLocation());
-    }
-
-    const tokensPayload = {
-      code: authCode,
-      state,
-    };
-
-    let tokenDecision;
-
-    try {
-      tokenDecision = await axios({
-        method: 'post',
-        url: tokenDecisionEndpoint,
-        headers: {
-          'Content-Type': 'application/json',
-        },
-        data: tokensPayload,
-      });
-    } catch (error) {
-      const err = error as AxiosError;
-      console.log(`Error while handling the redirect. Cause : ${err.message}`);
-
-      if (err.response !== undefined && err.response.status === 403) {
-        const response = {
-          title: 'Permission denied',
-          message: '',
-        };
-
-        if (err.response.data.hasOwnProperty('message')) {
-          response.message = err.response.data.message;
-        } else {
-          response.message = 'User is not allowed access the instance.';
+function setRoutes(app: Express, client: BaseClient, redirectUri: string): void {
+  // Initializing OAuth middleware.
+  app.use(oauthRouter(client, redirectUri, reduceRefreshDateSeconds));
+  // Authentication filter for API requests
+  app.use('/api', async (req: Request, resp: Response, next: NextFunction) => {
+    if (!isAuthenticated(req.session)) {
+      return next({ response: { status: 401 } });
+    } else {
+      const tokenSet = new TokenSet(req.session.tokenSet);
+      if (tokenSet.expired()) {
+        refreshPromises[req.session.id] ??= client.refresh(tokenSet).then((token) => {
```

Comment:
This is a very elegant solution. Only downside req.session.save can happen multiple times but this should not be a big issue. I think this is working fine if you have a single instance - one you have multiple instances(load balance) you might run into issues here.